aten/src/ATen/core/TensorImpl.h

```diff
@@ -220,5 +312,753 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
+        data_type_.copy()(src.raw_data(), raw_mutable_data(), size());
+      } else {
+        // We'll need to use a non-CPU context to perform the copy if
+        // one of the context is not CPU since only non-CPU context
+        // knows how to copy between CPU and that context
+        if (src.GetDeviceType() != ::at::DeviceType::CPU || GetDeviceType() == ::at::DeviceType::CPU) {
+          if (!context) {
+            src.CreateContext()->CopyBytesToDevice(
+                nbytes(), src.raw_data(), raw_mutable_data(), GetDeviceType());
+          } else {
+            CAFFE_ENFORCE(
+                context->device_type() == src.GetDeviceType(),
+                "Type for provided context does not match the type of source");
+            context->CopyBytesToDevice(
+                nbytes(), src.raw_data(), raw_mutable_data(), GetDeviceType());
+          }
+        } else {
+          // In case source context is CPU, and target context is non-CPU
+          // We'll have to create a Context from target and perform the
+          // copy using that context
+          CreateContext()->CopyBytesFromCPU(
+              nbytes(), src.raw_data(), raw_mutable_data());
+        }
+      }
+    }
+  }
+
+  /**
+   * @brief Extend the outer-most dimension of this tensor
+   *        to dimension of `num`.
+   */
+  void ExtendTo(int64_t num, float growthPct, at::BaseContext* context) {
+    CAFFE_ENFORCE_GE_WITH_CALLER(sizes_.size(), 1u);
+    CAFFE_ENFORCE_GE_WITH_CALLER(growthPct, 0);
+    CAFFE_ENFORCE(context != nullptr, "Context must be provided.");
+    Extend(num - sizes_[0], growthPct, context);
+  }
+
+  /**
+   * @brief Extends the outer-most dimension of this tensor by num elements,
+   * preserving the existing data.
+   *
+   * The underlying data may be reallocated in order to accommodate the new
+   * elements, in which case this tensors' capacity is grown at a factor of
+   * growthPct. This ensures that Extend runs on an amortized O(1) time
+   * complexity.
+   */
+  void Extend(int64_t num, float growthPct, at::BaseContext* context) {
+    CAFFE_ENFORCE_GE_WITH_CALLER(sizes_.size(), 1u);
+    CAFFE_ENFORCE_GE_WITH_CALLER(
+        num, 0, "`num` must be non-negative for Extend");
+    CAFFE_ENFORCE_WITH_CALLER(
+        is_contiguous_,
+        "Right now Extend is only supported for contiguous Tensor.");
+    auto newDims = sizes_;
+    newDims[0] += num;
+    if (!storage_.data()) {
+      Resize(newDims);
+      return;
+    }
+    auto newNumel = std::accumulate(
+        newDims.begin(),
+        newDims.end(),
+        static_cast<int64_t>(1),
+        std::multiplies<int64_t>());
+    if (newNumel * storage_.itemsize() <= storage_.capacity()) {
+      sizes_ = newDims;
+      numel_ = newNumel;
+      return;
+    }
+    auto newCapacity = sizes_;
+    newCapacity[0] = std::max<size_t>(
+        newDims[0], std::ceil(sizes_[0] * (growthPct + 100) / 100));
+    auto oldData = std::move(storage_.data_ptr());
```

Comment:
it's not this PR's issue, but we should refactor it, mutable reference getters are evil